### PR TITLE
chore: add more mobile apis

### PR DESCRIPTION
**Pull request by hanszhoou**
## Description

Closes # (NO_ISSUE)

## Type of change

- [ ] Documentation
- [ ] Code refactoring (Restructuring existing code w/o changing its observable behavior)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (a fix or feature that would make something no longer possible to do/require old user must upgrade their Mask Network to this new version)

## Checklist

- [ ] My code follows the style guidelines of this project.
- [ ] I have performed a self-review of my own code.
  - [ ] I have removed all in development `console.log`s
  - [ ] I have removed all commented code.
- [ ] I have commented on my code, particularly in hard-to-understand areas.
- [ ] I have read [Internationalization Guide](https://github.com/DimensionDev/Maskbook/blob/develop/docs/i18n-guide.md) and moved text fields to the i18n JSON file.

If this PR depends on external APIs:

- [ ] I have configured those APIs with CORS headers to let extension requests get passed. 
  - chrome extension: `chrome-extension://[id]`
  - firefox extension: `moz-extension://[id]`
- [ ] I have delegated all web requests to the background service via the internal RPC bridge.

### CODE REVIEW

File: packages/mask/background/database/avatar-cache/avatar.ts [hunk 1]
```diff
@@ -2,21 +2,27 @@ import { ProfileIdentifier } from '@masknet/shared-base'
...
 /**
- * Get a (cached) blob url for an identifier.
+ * Get a (cached) blob url for an identifier. No cache for native api.
  * ? Because of cross-origin restrictions, we cannot use blob url here. sad :(
  */
 export const queryAvatarDataURL = memoizePromise(
     async function (identifier: IdentifierWithAvatar): Promise<string | undefined> {
+        if (hasNativeAPI) {
+            const avatar = await nativeAPI?.api.query_avatar({ identifier: identifier.toText() })
+            if (!avatar) throw new Error('Avatar not found')
```
Comment (Jack-Works) [inline packages/mask/background/database/avatar-cache/avatar.ts]: why not `return undefined`?

File: packages/mask/background/database/avatar-cache/avatar.ts [hunk 2]
```diff
@@ -2,21 +2,27 @@ import { ProfileIdentifier } from '@masknet/shared-base'
...
+import { hasNativeAPI, nativeAPI } from '../../../shared/native-rpc'
 import { blobToDataURL } from '@dimensiondev/kit'
 import { createTransaction } from '../utils/openDB'
 
 /**
- * Get a (cached) blob url for an identifier.
+ * Get a (cached) blob url for an identifier. No cache for native api.
  * ? Because of cross-origin restrictions, we cannot use blob url here. sad :(
  */
 export const queryAvatarDataURL = memoizePromise(
```
Comment (Jack-Works) [inline packages/mask/background/database/avatar-cache/avatar.ts]: hasNativeAPI ? function (id) { ... } : memorizePromise(function (id) { ... })

File: packages/mask/background/database/avatar-cache/avatar.ts [hunk 3]
```diff
@@ -29,6 +35,10 @@ export const queryAvatarDataURL = memoizePromise(
 export async function storeAvatar(identifier: IdentifierWithAvatar, avatar: ArrayBuffer | string): Promise<void> {
     if (identifier instanceof ProfileIdentifier && identifier.isUnknown) return
     try {
+        if (hasNativeAPI) {
+            await nativeAPI?.api.store_avatar({ identifier: identifier.toText(), avatar: avatar as string })
```
Comment (Jack-Works) [inline packages/mask/background/database/avatar-cache/avatar.ts:43]: You didn't handle the ArrayBuffer type

Comment (jk234ert) [reply packages/mask/background/database/avatar-cache/avatar.ts:43]: `storeAvatar` used with ArrayBuffer is only used in the JS dashboard to set an avatar for the persona, while this is unreachable on our mobile app, we could just ignore the ArrayBuffer type param IMO.
Check the avatar type, do nothing with ArrayBuffer with some comments would be better here @zhouhanseng

File: packages/mask/background/database/post/app.ts [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+        postBy: record.postBy ? record.postBy.toText() : undefined,
+        identifier: record.identifier.toText(),
+        postCryptoKey: record.postCryptoKey,
+        recipients:
+            record.recipients === 'everyone'
+                ? 'everyone'
+                : record.recipients
+                ? Object.fromEntries(record.recipients.__raw_map__)
+                : undefined,
+        foundAt: record.foundAt ? record.foundAt.getTime() : undefined,
```
Comment (septs) [inline packages/mask/background/database/post/app.ts]:
```suggestion
        foundAt: record.foundAt?.getTime(),
```

File: packages/mask/background/database/post/app.ts [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+        encryptBy: record.encryptBy?.toText(),
+        url: record.url,
+        summary: record.summary,
+        interestedMeta: record.interestedMeta,
+    } as NativePostRecord
```
Comment (septs) [inline packages/mask/background/database/post/app.ts]:
```suggestion
    }
```

File: packages/mask/background/database/post/app.ts [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+        recipients:
+            record.recipients === 'everyone'
+                ? 'everyone'
+                : new IdentifierMap(new Map(Object.entries(record.recipients))),
+        foundAt: new Date(record.foundAt),
+        encryptBy: record.encryptBy ? Identifier.fromString(record.encryptBy).unwrap() : undefined,
+        url: record.url,
+        summary: record.summary,
+        interestedMeta: record.interestedMeta,
+    } as PostRecord
```
Comment (septs) [inline packages/mask/background/database/post/app.ts]:
```suggestion
    }
```

File: packages/mask/background/database/post/app.ts [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+function postInNative(record: Partial<PostRecord> & Pick<PostRecord, 'identifier'>): Partial<NativePostRecord> {
+    return {
+        postBy: record.postBy ? record.postBy.toText() : undefined,
+        identifier: record.identifier.toText(),
+        postCryptoKey: record.postCryptoKey,
+        recipients:
+            record.recipients === 'everyone'
+                ? Object.fromEntries(new Map())
+                : record.recipients
+                ? Array.from(record.recipients).map(
```
Comment (sxiaojian88) [inline packages/mask/background/database/post/app.ts:78]: Recipients is a empty map when`record.recipients === 'everyone'`, and an array when `record.recipients != 'everyone'`.
Please always use map by remove `Array.from(record.recipients)`

Comment (sxiaojian88) [reply packages/mask/background/database/post/app.ts:78]: Screen Shot 2022-02-22 at 3 34 11 PM
Please use the original format

**Comment by sxiaojian88** [comment]
Post query create update and merge have been tested,
Avatar update has been tested, iOS client can get avatar.
Avatar query hasn't been tested, I don't know where to display avatar in web page.
Look good to me.

**Comment by zhouhanseng** [comment]
Native side colleague confirmed.

**Comment by jk234ert** [comment]
> Avatar query hasn't been tested, I don't know where to display avatar in web page.

It's in the dashboard of our extension, which is hidden on the app.

**Review by BillySh1 — APPROVED** [review]

**Review by sxiaojian88 — APPROVED** [review]

File: packages/public-api/src/mobile.ts [hunk 8]
```diff
@@ -100,6 +101,17 @@ export interface SharedNativeAPIs {
...
+    store_avatar(params: { identifier: string; avatar: string }): Promise<void>
+    create_post(params: { post: PostRecord }): Promise<PostRecord>
+    query_post(params: { identifier: string }): Promise<PostRecord>
+    query_posts(params: {
+        encryptBy?: string
+        userIds: string[]
+        network?: string
+        pageOption?: PageOption
+    }): Promise<PostRecord[]>
+    update_post(params: { post: Partial<PostRecord>; options: { mode: 0 | 1 } }): Promise<PostRecord[]>
```
Comment (Jack-Works) [inline packages/public-api/src/mobile.ts:114]: please document what 0 and 1 means.

**Pull request merged**